File: bigbluebutton-apps/src/main/scala/org/bigbluebutton/core/apps/presentation/red5/PresentationClientMessageSender.scala
package org.bigbluebutton.core.apps.presentation.red5

import org.bigbluebutton.conference.meeting.messaging.red5.ConnectionInvokerService
import org.bigbluebutton.core.api._
import org.bigbluebutton.conference.meeting.messaging.red5.BroadcastClientMessage
import org.bigbluebutton.conference.meeting.messaging.red5.DirectClientMessage
import collection.JavaConversions._
import com.google.gson.Gson
import java.util.ArrayList
import org.bigbluebutton.core.apps.presentation.Page
import org.bigbluebutton.core.apps.presentation.Presentation


class PresentationClientMessageSender(service: ConnectionInvokerService) extends OutMessageListener2 {
  
	private val OFFICE_DOC_CONVERSION_SUCCESS_KEY = "OFFICE_DOC_CONVERSION_SUCCESS";
    private val OFFICE_DOC_CONVERSION_FAILED_KEY = "OFFICE_DOC_CONVERSION_FAILED";
    private val SUPPORTED_DOCUMENT_KEY = "SUPPORTED_DOCUMENT";
    private val UNSUPPORTED_DOCUMENT_KEY = "UNSUPPORTED_DOCUMENT";
    private val PAGE_COUNT_FAILED_KEY = "PAGE_COUNT_FAILED";
    private val PAGE_COUNT_EXCEEDED_KEY = "PAGE_COUNT_EXCEEDED";	
    private val GENERATED_SLIDE_KEY = "GENERATED_SLIDE";
    private val GENERATING_THUMBNAIL_KEY = "GENERATING_THUMBNAIL";
    private val GENERATED_THUMBNAIL_KEY = "GENERATED_THUMBNAIL";
    private val CONVERSION_COMPLETED_KEY = "CONVERSION_COMPLETED";
    
  def handleMessage(msg: IOutMessage) {
    msg match {
      case msg: ClearPresentationOutMsg             => handleClearPresentationOutMsg(msg)
//      case msg: RemovePresentationOutMsg            => handleRemovePresentationOutMsg(msg)
      case msg: GetPresentationInfoOutMsg           => handleGetPresentationInfoOutMsg(msg)
      case msg: SendCursorUpdateOutMsg              => handleSendCursorUpdateOutMsg(msg)
      case msg: ResizeAndMoveSlideOutMsg            => handleResizeAndMoveSlideOutMsg(msg)
      case msg: GotoSlideOutMsg                     => handleGotoSlideOutMsg(msg)
      case msg: SharePresentationOutMsg             => handleSharePresentationOutMsg(msg)
      case msg: GetSlideInfoOutMsg                  => handleGetSlideInfoOutMsg(msg)
      case msg: PresentationConversionProgress      => handlePresentationConversionProgress(msg)
      case msg: PresentationConversionError         => handlePresentationConversionError(msg)
      case msg: PresentationPageGenerated           => handlePresentationPageGenerated(msg)
      case msg: PresentationConversionDone          => handlePresentationConversionDone(msg)
      case _ => // do nothing
    }
  }
  
  private def handleClearPresentationOutMsg(msg: ClearPresentationOutMsg) {
      
  }
  
  private def handlePresentationConversionProgress(msg: PresentationConversionProgress) {
	val args = new java.util.HashMap[String, String]();
	args.put("meetingID", msg.meetingID);
	args.put("code", msg.code);
	args.put("presentationID", msg.presentationId);
	args.put("presentationName", msg.presentationName);
	args.put("messageKey", msg.messageKey);
	
	val message = new java.util.HashMap[String, Object]() 
	val gson = new Gson();
  	message.put("msg", gson.toJson(args))
  	
//  	println("PresentationClientMessageSender - handlePresentationConversionProgress \n" + message.get("msg") + "\n")

    val m = new BroadcastClientMessage(msg.meetingID, "conversionUpdateMessageCallback", message);
	service.sendMessage(m);	
  }

  private def handlePresentationConversionError(msg: PresentationConversionError) {
    val args = new java.util.HashMap[String, String]();
	args.put("meetingID", msg.meetingID);
	args.put("code", msg.code);
	args.put("presentationID", msg.presentationId);
	args.put("presentationName", msg.presentationName);
	args.put("messageKey", msg.messageKey);
	args.put("numberOfPages", msg.numberOfPages.toString);
	args.put("maxNumberPages", msg.maxNumberPages.toString);

	val message = new java.util.HashMap[String, Object]() 
	val gson = new Gson();
  	message.put("msg", gson.toJson(args))
  	
//  	println("PresentationClientMessageSender - handlePresentationConversionError \n" + message.get("msg") + "\n")

	val m = new BroadcastClientMessage(msg.meetingID, "pageCountExceededUpdateMessageCallback", message);
    service.sendMessage(m);    
  }

  private def handlePresentationPageGenerated(msg: PresentationPageGenerated) {
    val args = new java.util.HashMap[String, String]();
	args.put("meetingID", msg.meetingID);
	args.put("code", msg.code);
	args.put("presentationID", msg.presentationId);
	args.put("presentationName", msg.presentationName);
	args.put("messageKey", msg.messageKey);
	args.put("numberOfPages", msg.numberOfPages.toString);
	args.put("pagesCompleted", msg.pagesCompleted.toString);

	val message = new java.util.HashMap[String, Object]() 
	val gson = new Gson();
  	message.put("msg", gson.toJson(args))
  	
//  	println("PresentationClientMessageSender - handlePresentationPageGenerated \n" + message.get("msg") + "\n")

	val m = new BroadcastClientMessage(msg.meetingID, "generatedSlideUpdateMessageCallback", message);
    service.sendMessage(m);    
  }
  
  private def handlePresentationConversionDone(msg: PresentationConversionDone) {
    val args = new java.util.HashMap[String, Object]()
	args.put("meetingID", msg.meetingID);
	args.put("code", msg.code);
	
	val presentation = new java.util.HashMap[String, Object]();
	presentation.put("id", msg.presentation.id)
	presentation.put("name", msg.presentation.name)
	presentation.put("current", msg.presentation.current:java.lang.Boolean)
	
	val pages = new ArrayList[Page]()
	
	msg.presentation.pages.values foreach {p =>
//      println("PresentationClientMessageSender **** Page [" + p.id + "," + p.num + "]")
      pages.add(p)
    }
	
	presentation.put("pages", pages)
	
	args.put("presentation", presentation);

	val message = new java.util.HashMap[String, Object]() 
	val gson = new Gson();
  	message.put("msg", gson.toJson(args))
  		
//  	println("PresentationClientMessageSender - handlePresentationConversionDone \n" + message.get("msg") + "\n")

	val m = new BroadcastClientMessage(msg.meetingID, "conversionCompletedUpdateMessageCallback", message);
    service.sendMessage(m);      
  }
 
  private def handleRemovePresentationOutMsg(msg: RemovePresentationOutMsg) {
	val args = new java.util.HashMap[String, Object]();
	args.put("presentationID", msg.presentationID);

	val message = new java.util.HashMap[String, Object]() 
	val gson = new Gson();
  	message.put("msg", gson.toJson(args))
  	
	val m = new BroadcastClientMessage(msg.meetingID, "removePresentationCallback", message);
	service.sendMessage(m);    
  }
  
  private def handleGetPresentationInfoOutMsg(msg: GetPresentationInfoOutMsg) {
    val info = msg.info
    
    // Build JSON
    val args = new java.util.HashMap[String, Object]()
	  args.put("meetingID", msg.meetingID);
	
    // Create a map for our current presenter
    val presenter = new java.util.HashMap[String, String]()
    presenter.put("userId", info.presenter.userId)
    presenter.put("name", info.presenter.name)
    presenter.put("assignedBy", info.presenter.assignedBy)
	
    args.put("presenter", presenter)
    
    // Create an array for our presentations
    val presentations = new ArrayList[java.util.HashMap[String, Object]]     
    info.presentations.foreach { pres =>
	   val presentation = new java.util.HashMap[String, Object]();
	   presentation.put("id", pres.id)
	   presentation.put("name", pres.name)
	   presentation.put("current", pres.current:java.lang.Boolean)      
	   
	   // Get the pages for a presentation
       val pages = new ArrayList[Page]()	
	   pres.pages.values foreach {p =>
         pages.add(p)
       }   
	  // store the pages in the presentation 
	  presentation.put("pages", pages)
	
	  // add this presentation into our presentations list
	  presentations.add(presentation);	   
    }
    
    // add the presentation to our map to complete our json
    args.put("presentations", presentations)

    val message = new java.util.HashMap[String, Object]() 
	  val gson = new Gson();
  	message.put("msg", gson.toJson(args))
  	 	
  	println("***** PresentationClientMessageSender - handleGetPresentationInfoOutMsg to user[" +msg.requesterID + "] message[" + message.get("msg") + "]")
  	
	val m = new DirectClientMessage(msg.meetingID, msg.requesterID, "getPresentationInfoReply", message);
//	service.sendMessage(m);	
  }
  
  private def handleSendCursorUpdateOutMsg(msg: SendCursorUpdateOutMsg) {
	val args = new java.util.HashMap[String, Object]();
	args.put("xPercent", msg.xPercent:java.lang.Double);
	args.put("yPercent", msg.yPercent:java.lang.Double);

	val message = new java.util.HashMap[String, Object]() 
	val gson = new Gson();
  	message.put("msg", gson.toJson(args))
  	  	
	val m = new BroadcastClientMessage(msg.meetingID, "PresentationCursorUpdateCommand", message);
	service.sendMessage(m);	    
  }
  
  private def handleResizeAndMoveSlideOutMsg(msg: ResizeAndMoveSlideOutMsg) {
	val args = new java.util.HashMap[String, Object]();
	args.put("id", msg.page.id)
	args.put("num", msg.page.num:java.lang.Integer)
	args.put("current", msg.page.current:java.lang.Boolean)
	args.put("swfUri", msg.page.swfUri)
	args.put("txtUri", msg.page.txtUri)
	args.put("pngUri", msg.page.pngUri)
	args.put("thumbUri", msg.page.thumbUri)
	args.put("xOffset", msg.page.xOffset:java.lang.Double);
	args.put("yOffset", msg.page.yOffset:java.lang.Double);
	args.put("widthRatio", msg.page.widthRatio:java.lang.Double);
	args.put("heightRatio", msg.page.heightRatio:java.lang.Double);

	val message = new java.util.HashMap[String, Object]() 
	val gson = new Gson();
  	message.put("msg", gson.toJson(args))
  	  	
	val m = new BroadcastClientMessage(msg.meetingID, "moveCallback", message);
	service.sendMessage(m);	    
  }
  
  private def handleGotoSlideOutMsg(msg: GotoSlideOutMsg) {
	val args = new java.util.HashMap[String, Object]();
	args.put("id", msg.page.id)
	args.put("num", msg.page.num:java.lang.Integer)
	args.put("current", msg.page.current:java.lang.Boolean)
	args.put("swfUri", msg.page.swfUri)
	args.put("txtUri", msg.page.txtUri)
	args.put("pngUri", msg.page.pngUri)
	args.put("thumbUri", msg.page.thumbUri)
	args.put("xOffset", msg.page.xOffset:java.lang.Double);
	args.put("yOffset", msg.page.yOffset:java.lang.Double);
	args.put("widthRatio", msg.page.widthRatio:java.lang.Double);
	args.put("heightRatio", msg.page.heightRatio:java.lang.Double);

	val message = new java.util.HashMap[String, Object]() 
	val gson = new Gson();
  	message.put("msg", gson.toJson(args))
  		
	val m = new BroadcastClientMessage(msg.meetingID, "goToSlideCallback", message);
	service.sendMessage(m);	    
  }
  
  private def handleSharePresentationOutMsg(msg: SharePresentationOutMsg) {
	val args = new java.util.HashMap[String, Object]();
	
	val presentation = new java.util.HashMap[String, Object]();
	presentation.put("id", msg.presentation.id)
	presentation.put("name", msg.presentation.name)
	presentation.put("current", msg.presentation.current:java.lang.Boolean)      
	   
	// Get the pages for a presentation
    val pages = new ArrayList[Page]()	
	msg.presentation.pages.values foreach {p =>
     pages.add(p)
    }   
	// store the pages in the presentation 
	presentation.put("pages", pages)
	
	args.put("presentation", presentation);
	
    val message = new java.util.HashMap[String, Object]() 
	val gson = new Gson();
  	message.put("msg", gson.toJson(args))
  	  	
	val m = new BroadcastClientMessage(msg.meetingID, "sharePresentationCallback", message);
	service.sendMessage(m);	    
  }
  
  private def handleGetSlideInfoOutMsg(msg: GetSlideInfoOutMsg) {
    val args = new java.util.HashMap[String, Object]();
    args.put("xOffset", msg.page.xOffset:java.lang.Double);
    args.put("yOffest", msg.page.yOffset:java.lang.Double);
    args.put("widthRatio", msg.page.widthRatio:java.lang.Double);
    args.put("heightRatio", msg.page.heightRatio:java.lang.Double);	

    val message = new java.util.HashMap[String, Object]() 
    val gson = new Gson();
    message.put("msg", gson.toJson(args))
  	
    val m = new DirectClientMessage(msg.meetingID, msg.requesterID, "getPresentationInfoReply", message);
    service.sendMessage(m);    
  }
}

File: bigbluebutton-apps/src/main/java/org/bigbluebutton/red5/client/PresentationClientMessageSender.java
package org.bigbluebutton.red5.client;

import java.util.HashMap;
import java.util.Map;

import org.bigbluebutton.common.messages.Constants;
import org.bigbluebutton.common.messages.GetPresentationInfoReplyMessage;
import org.bigbluebutton.common.messages.GetSlideInfoReplyMessage;
import org.bigbluebutton.common.messages.GoToSlideReplyMessage;
import org.bigbluebutton.common.messages.PresentationConversionDoneMessage;
import org.bigbluebutton.common.messages.PresentationConversionErrorMessage;
import org.bigbluebutton.common.messages.PresentationConversionProgressMessage;
import org.bigbluebutton.common.messages.PresentationCursorUpdateMessage;
import org.bigbluebutton.common.messages.PresentationPageGeneratedReplyMessage;
import org.bigbluebutton.common.messages.PresentationPageResizedMessage;
import org.bigbluebutton.common.messages.PresentationRemovedMessage;
import org.bigbluebutton.common.messages.PresentationSharedMessage;
import org.bigbluebutton.red5.client.messaging.BroadcastClientMessage;
import org.bigbluebutton.red5.client.messaging.ConnectionInvokerService;
import org.bigbluebutton.red5.client.messaging.DirectClientMessage;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

public class PresentationClientMessageSender {
	private ConnectionInvokerService service;
	
	public PresentationClientMessageSender(ConnectionInvokerService service) {
		this.service = service;
	}
	
	public void handlePresentationMessage(String message) {
		JsonParser parser = new JsonParser();
		JsonObject obj = (JsonObject) parser.parse(message);
		
		if (obj.has("header") && obj.has("payload")) {
			JsonObject header = (JsonObject) obj.get("header");

			if (header.has("name")) {
				String messageName = header.get("name").getAsString();
				switch (messageName) {
				  case PresentationRemovedMessage.PRESENTATION_REMOVED:
					  processPresentationRemovedMessage(message);
					  break;
				  case GetPresentationInfoReplyMessage.GET_PRESENTATION_INFO_REPLY:
					  processGetPresentationInfoReplyMessage(message);
					  break;
				  case GoToSlideReplyMessage.GO_TO_SLIDE_REPLY:
					  processGoToSlideMessage(message);
					  break;
				  case PresentationConversionProgressMessage.PRESENTATION_CONVERSION_PROGRESS:
					  processPresentationConversionProgress(message);
					  break;
				  case PresentationPageGeneratedReplyMessage.PRESENTATION_PAGE_GENERATED:
					  processPresentationPageGeneratedReply(message);
					  break;
				  case PresentationCursorUpdateMessage.PRESENTATION_CURSOR_UPDATED:
					  processPresentationCursorUpdate(message);
					  break;
				  case PresentationConversionErrorMessage.PRESENTATION_CONVERSION_ERROR:
					  processPresentationConversionError(message);
					  break;
				  case GetSlideInfoReplyMessage.GET_SLIDE_INFO:
					  processGetSlideInfoReply(message);
					  break;
				  case PresentationConversionDoneMessage.PRESENTATION_CONVERSION_DONE:
					  processPresentationConversionDone(message);
					  break;
				  case PresentationPageResizedMessage.PRESENTATION_PAGE_RESIZED:
					  processPresentationPageResized(message);
					  break;
				  case PresentationSharedMessage.PRESENTATION_SHARED_MESSAGE:
					  processPresentationSharedMessage(message);
					  break;
				}
			}
		}		
	}

	private void processPresentationRemovedMessage(String json) {
		PresentationRemovedMessage msg = PresentationRemovedMessage.fromJson(json);
		if (msg != null) {
			Map<String, Object> args = new HashMap<String, Object>();  
			args.put("presentationID", msg.presentationId);
			  
			Map<String, Object> message = new HashMap<String, Object>();
			Gson gson = new Gson();
			message.put("msg", gson.toJson(args));
			  	      	    
			BroadcastClientMessage m = new BroadcastClientMessage(msg.meetingId, "removePresentationCallback", message);
			service.sendMessage(m);		
		}
	}
	
	private void processGetPresentationInfoReplyMessage(String json) {
		GetPresentationInfoReplyMessage msg = GetPresentationInfoReplyMessage.fromJson(json);
		if (msg != null) {
			Map<String, Object> args = new HashMap<String, Object>();  
			args.put("meetingID", msg.meetingId);			
		    args.put("presenter", msg.presenter);
		    args.put("presentations", msg.presentations);
		    		    
			Map<String, Object> message = new HashMap<String, Object>();
			Gson gson = new Gson();
			message.put("msg", gson.toJson(args));
			  	    	  	    
			DirectClientMessage m = new DirectClientMessage(msg.meetingId, msg.requesterId, "getPresentationInfoReply", message);
			service.sendMessage(m);		
		}		
	}
	
	private void processPresentationSharedMessage(String json) {

		PresentationSharedMessage msg = PresentationSharedMessage.fromJson(json);
		if (msg != null) {

			Map<String, Object> presentation = new HashMap<String, Object>();
			presentation.put("id", msg.presentation.get("id"));
			presentation.put("name", msg.presentation.get("name"));
			presentation.put("current", msg.presentation.get("current"));
			presentation.put("pages", msg.presentation.get("pages"));

			Map<String, Object> args = new HashMap<String, Object>();
			args.put("presentation", presentation);

			Map<String, Object> message = new HashMap<String, Object>();
			Gson gson = new Gson();
			message.put("msg", gson.toJson(args));

			BroadcastClientMessage m = new BroadcastClientMessage(msg.meetingId, "sharePresentationCallback", message);
			service.sendMessage(m);
		}
	}
	
	private void processPresentationPageResized(String json) {

		PresentationPageResizedMessage msg = PresentationPageResizedMessage.fromJson(json);
		if (msg != null) {
			Map<String, Object> args = new HashMap<String, Object>();

			args.put("id", msg.page.get("id"));
			args.put("num", msg.page.get("num"));
			args.put("current", msg.page.get("current"));
			args.put("swfUri", msg.page.get("swf_uri"));
			args.put("txtUri", msg.page.get("txt_uri"));
			args.put("svgUri", msg.page.get("svg_uri"));
			args.put("thumbUri", msg.page.get("thumb_uri"));
			args.put("xOffset", msg.page.get("x_offset"));
			args.put("yOffset", msg.page.get("y_offset"));
			args.put("widthRatio", msg.page.get("width_ratio"));
			args.put("heightRatio", msg.page.get("height_ratio"));

			Map<String, Object> message = new HashMap<String, Object>();
			Gson gson = new Gson();
			message.put("msg", gson.toJson(args));

			BroadcastClientMessage m = new BroadcastClientMessage(msg.meetingId, "moveCallback", message);
			service.sendMessage(m);
		}
	}
	
	private void processPresentationConversionDone(String json) {

		PresentationConversionDoneMessage msg = PresentationConversionDoneMessage.fromJson(json);
		if (msg != null) {
			Map<String, Object> args = new HashMap<String, Object>();
			args.put("meetingID", msg.meetingId);
			args.put("code", msg.code);

			Map<String, Object> presentation = new HashMap<String, Object>();
			presentation.put("id", msg.presentation.get("id"));
			presentation.put("name", msg.presentation.get("name"));
			presentation.put("current", msg.presentation.get("current"));
			presentation.put("pages", msg.presentation.get("pages"));

			args.put("presentation", presentation);

			Map<String, Object> message = new HashMap<String, Object>();
			Gson gson = new Gson();
			message.put("msg", gson.toJson(args));

			BroadcastClientMessage m = new BroadcastClientMessage(msg.meetingId, "conversionCompletedUpdateMessageCallback", message);
			service.sendMessage(m);
		}
	}

	private void processGetSlideInfoReply(String json) {
		GetSlideInfoReplyMessage msg = GetSlideInfoReplyMessage.fromJson(json);
		if (msg != null) {
			Map<String, Object> args = new HashMap<String, Object>();
			args.put("meetingID", msg.meetingId);
			args.put("xOffset", msg.xOffset);
			args.put("yOffest", msg.yOffset);
			args.put("widthRatio", msg.widthRatio);
			args.put("heightRatio", msg.heightRatio);

			Map<String, Object> message = new HashMap<String, Object>();
			Gson gson = new Gson();
			message.put("msg", gson.toJson(args));

			DirectClientMessage m = new DirectClientMessage(msg.meetingId, msg.requesterId, "getSlideInfoReply", message);
			service.sendMessage(m);
		}
	}

	private void processPresentationConversionError(String json) {
		PresentationConversionErrorMessage msg = PresentationConversionErrorMessage.fromJson(json);
		if (msg != null) {
			Map<String, Object> args = new HashMap<String, Object>();
			args.put("meetingID", msg.meetingId);
			args.put("code", msg.code);
			args.put("presentationID", msg.presentationId);
			args.put("presentationName", msg.presentationName);
			args.put("messageKey", msg.messageKey);
			args.put("numberOfPages", msg.numPages);
			args.put("maxNumberPages", msg.maxNumPages);
			
			Map<String, Object> message = new HashMap<String, Object>();
			Gson gson = new Gson();
			message.put("msg", gson.toJson(args));
			
			BroadcastClientMessage m = new BroadcastClientMessage(msg.meetingId, "pageCountExceededUpdateMessageCallback", message);
			service.sendMessage(m);
		}
	}

	private void processPresentationCursorUpdate(String json) {
		PresentationCursorUpdateMessage msg = PresentationCursorUpdateMessage.fromJson(json);
		if (msg != null) {
			Map<String, Object> args = new HashMap<String, Object>();
			args.put("meetingID", msg.meetingId);
			args.put("xPercent", msg.xPercent);
			args.put("yPercent", msg.yPercent);

			Map<String, Object> message = new HashMap<String, Object>();
			Gson gson = new Gson();
			message.put("msg", gson.toJson(args));

			BroadcastClientMessage m = new BroadcastClientMessage(msg.meetingId, "PresentationCursorUpdateCommand", message);
			service.sendMessage(m);
		}
	}

	private void processPresentationPageGeneratedReply(String json) {
		PresentationPageGeneratedReplyMessage msg = PresentationPageGeneratedReplyMessage.fromJson(json);
		if (msg != null) {
			Map<String, Object> args = new HashMap<String, Object>();
			args.put("presentationID", msg.presentationId);
			args.put("meetingID", msg.meetingId);
			args.put("code", msg.code);
			args.put("messageKey", msg.messageKey);
			args.put("presentationName", msg.presentationName);
			args.put("pagesCompleted", msg.pagesCompleted);
			args.put("numberOfPages", msg.numPages);

			Map<String, Object> message = new HashMap<String, Object>();
			Gson gson = new Gson();
			message.put("msg", gson.toJson(args));

			BroadcastClientMessage m = new BroadcastClientMessage(msg.meetingId, "generatedSlideUpdateMessageCallback", message);
			service.sendMessage(m);
		}
	}

	private void processPresentationConversionProgress(String json) {
		PresentationConversionProgressMessage msg = PresentationConversionProgressMessage.fromJson(json);
		if (msg != null) {
			Map<String, Object> args = new HashMap<String, Object>();
			args.put("presentationID", msg.presentationId);
			args.put("meetingID", msg.meetingId);
			args.put("code", msg.code);
			args.put("messageKey", msg.messageKey);
			args.put("presentationName", msg.presentationName);

			Map<String, Object> message = new HashMap<String, Object>();
			Gson gson = new Gson();
			message.put("msg", gson.toJson(args));

			BroadcastClientMessage m = new BroadcastClientMessage(msg.meetingId, "conversionUpdateMessageCallback", message);
			service.sendMessage(m);
		}
	}


	private void processGoToSlideMessage(String json) {
		GoToSlideReplyMessage msg = GoToSlideReplyMessage.fromJson(json);
		if (msg != null) {
			Map<String, Object> args = new HashMap<String, Object>();
			args.put("id", msg.page.get(Constants.ID));
			args.put("widthRatio", msg.page.get(Constants.WIDTH_RATIO));
			args.put("yOffset", msg.page.get(Constants.Y_OFFSET));
			args.put("xOffset", msg.page.get(Constants.X_OFFSET));
			args.put("num", msg.page.get("num"));
			args.put("heightRatio", msg.page.get(Constants.HEIGHT_RATIO));
			args.put("svgUri", msg.page.get("svg_uri"));
			args.put("thumbUri", msg.page.get("thumb_uri"));
			args.put(Constants.CURRENT, msg.page.get(Constants.CURRENT));
			args.put("txtUri", msg.page.get("txt_uri"));
			args.put("swfUri", msg.page.get("swf_uri"));

			Map<String, Object> message = new HashMap<String, Object>();
			Gson gson = new Gson();
			message.put("msg", gson.toJson(args));

			BroadcastClientMessage m = new BroadcastClientMessage(msg.meetingId, "goToSlideCallback", message);
			service.sendMessage(m);
		}
	}	

}
